test/model/message_list_test.dart

```diff
@@ -0,0 +1,103 @@
+import 'package:zulip/model/store.dart';
+import '../api/fake_api.dart';
+import '../model/binding.dart';
+import '../model/test_store.dart';
+import '../example_data.dart' as eg;
+
+const int userId = 1;
+const int streamId = 2;
+
+Future<PerAccountStore> setupStore(ZulipStream stream) async {
+  await TestZulipBinding.instance.globalStore.add(eg.selfAccount, eg.initialSnapshot());
+  PerAccountStore store = await TestZulipBinding.instance.globalStore.perAccount(eg.selfAccount.id);
+  store.addUser(eg.user(userId: userId));
+  store.addStream(stream);
+  return store;
+}
+
+void main() async {
+  TestZulipBinding.ensureInitialized();
+  const narrow = StreamNarrow(streamId);
+
+  final ZulipStream stream = eg.stream(streamId: streamId);
+  PerAccountStore store = await setupStore(stream);
+
+  group('update message tests', () {
+
+    test('find message in message list returns index of message', () async {
+      MessageListView messageList = MessageListView.init(store: store, narrow: narrow);
+
+      final connection = store.connection as FakeApiConnection;
+      Message m1 = eg.streamMessage(id: 792, stream: stream);
+      Message m2 = eg.streamMessage(id: 793, stream: stream);
+      Message m3 = eg.streamMessage(id: 794, stream: stream);
+
+      connection.prepare(json: GetMessagesResult(
+        anchor: m1.id,
+        foundNewest: true,
+        foundOldest: true,
+        foundAnchor: true,
+        historyLimited: false,
+        messages: [m1, m2, m3],
+      ).toJson());
+
+      await messageList.fetch();
+
+      check(messageList.messages.length).equals(3);
+
+      int idx = messageList.findMessageWithId(793);
+      check(idx).equals(1);
+
+      idx = messageList.findMessageWithId(999);
+      check(idx).equals(-1);
+    });
+
+    test('update events are correctly applied to message', () async {
+      MessageListView messageList = MessageListView.init(store: store, narrow: narrow);
+
+      final connection = store.connection as FakeApiConnection;
+      String oldContent = "<p>Hello, world</p>";
+      String newContent = "<p>Hello, edited</p>";
+      int newTimestamp = 99999;
+
+      Message m = eg.streamMessage(id: 243, stream: stream, content: oldContent);
+      connection.prepare(json: GetMessagesResult(
+        anchor: m.id,
+        foundNewest: true,
+        foundOldest: true,
+        foundAnchor: true,
+        historyLimited: false,
+        messages: [m],
+      ).toJson());
+
+      await messageList.fetch();
+
+      UpdateMessageEvent updateEvent = UpdateMessageEvent(
+        id: 1,
+        messageId: m.id,
+        messageIds: [m.id],
+        flags: m.flags,
+        renderedContent: newContent,
+        editTimestamp: newTimestamp
+      );
+
+      Message oldMessage = messageList.messages[0];
+      check(oldMessage.content).equals(oldContent);
+
+      messageList.maybeUpdateMessage(updateEvent);
+
+      Message updatedMessage = messageList.messages[0];
```

Comment:
Good idea, will fix! 

Just to get into the weeds a bit - I think I got a little turned around here  because I was originally calling `maybeAddMessage` instead of mocking my messages out with `fetch`. So one thing that's a little tricky about these tests is that `oldMessage === newMessage` but `m !=== oldMessage`. I would still sort of assume that somewhere in the code that mocks out the API response, that it would still hold onto the original `Message m`. But perhaps it serializes/deserializes these objects when it sets up the mock response?

Reply:
Also, fixed.

Reply:
> So one thing that's a little tricky about these tests is that `oldMessage === newMessage` but `m !=== oldMessage`. I would still sort of assume that somewhere in the code that mocks out the API response, that it would still hold onto the original `Message m`. But perhaps it serializes/deserializes these objects when it sets up the mock response?

Ah. Yeah, here's the docs on `FakeApiConnection.prepare`:
```
  /// Prepare the response for the next request.
  ///
  /// […] The body of the response
  /// will be `body` if non-null, or `jsonEncode(json)` if `json` is non-null,
  /// or else ''.  The `body` and `json` parameters must not both be non-null.
```

So the `json` parameter is really just a shorthand for passing `body` with the string you'd get from serializing the JSON. (For that matter, the implementation shows that even `body` is a bit of a shorthand for passing the *bytes* you'd get from encoding the string as UTF-8; we just haven't yet had cause to write a test where the response bytes aren't valid UTF-8.)

Fundamentally that's because the fake is happening at the level of the HTTP client library, so it doesn't know anything about the structures that the endpoint binding code is going to want to return. This makes it useful for testing the endpoint bindings themselves (and indeed we use the same `prepare` function for those tests), and also makes it conveniently polymorphic for setting up data like this for whatever arbitrary endpoint the code under test is expected to invoke.